File: tools/capture_swarm_dpi.js
// capture_swarm_dpi.js — log every HID feature report Swarm II sends/reads through KONE_XP_AIR.dll.
// Attach to Swarm's process, then change the DPI in Swarm and click apply; every 0x06 frame Swarm
// writes prints as SEND, every profile read as GET. This is what we need to reproduce the DPI write.
var kone = Process.getModuleByName('KONE_XP_AIR.dll');
var sendAddr = null, getAddr = null;
kone.enumerateExports().forEach(function (e) {
    if (e.name === 'hid_send_feature_report') sendAddr = e.address;
    if (e.name === 'hid_get_feature_report') getAddr = e.address;
});

function hex(ptr, n) {
    if (n > 30) n = 30;                 // feature reports are 30 bytes on this device
    var out = [];
    for (var i = 0; i < n; i++) {
        var b = ptr.add(i).readU8().toString(16);
        out.push(b.length < 2 ? '0' + b : b);
    }
    return out.join(' ');
}

if (!sendAddr) {
    send({ line: 'ERROR: hid_send_feature_report not found in KONE_XP_AIR.dll' });
} else {
    // int hid_send_feature_report(hid_device *dev, const unsigned char *data, size_t length)
    // Show ONLY the profile-write / switch commands so the DPI sequence is short and unambiguous:
    //   0x45 profile-select, 0x46 profile block (page selects, writes, reads, commit), 0x47 buttons,
    //   0x49 commit, 0x4e activate.  Hidden: 0x44 status polling, 0x4d lighting, 06 00 receiver pings,
    //   and all GET responses (read REQUESTS still show as 06 01 46 07 sends).
    var KEEP = [0x45, 0x46, 0x47, 0x49, 0x4e];
    Interceptor.attach(sendAddr, {
        onEnter: function (args) {
            var b1 = args[1].add(1).readU8(), b2 = args[1].add(2).readU8();
            if (b1 !== 0x01 || KEEP.indexOf(b2) < 0) return;
            var len = args[2].toInt32();
            send({ dir: 'SEND', len: len, data: hex(args[1], len) });
        }
    });
    send({ line: 'hooked KONE_XP_AIR.dll — change the DPI in Swarm II now and click apply' });
}

File: tools/capture_swarm_dpi.py
"""capture_swarm_dpi.py — record the HID feature reports Swarm II sends when you change the DPI.

Windows only, Swarm II running. This is how we reproduce Swarm's real DPI write (our own 0x46 write is
rejected by the mouse; the .dat gives the block content but not the wire write sequence).

    python tools/capture_swarm_dpi.py
    python tools/capture_swarm_dpi.py --process "Turtle Beach Device Service.exe"   # if the UI exe shows nothing

Steps:
  1. Open Swarm II (so it owns the mouse).
  2. Run this. It attaches and prints "hooked ...".
  3. In Swarm II change the DPI to a distinctive value (e.g. 800 -> 3000) and click apply/save.
  4. Watch the SEND lines scroll. Press ENTER here to stop.
  5. Copy ALL the SEND/GET lines from step 3 and send them back.

If nothing prints when you change the DPI, the write is happening in the other process — re-run with
--process "Turtle Beach Device Service.exe". Run the terminal as Administrator if attach fails.
"""
import argparse
import os
import sys

import _common  # noqa: F401  (adds repo root to sys.path)
import frida

JS = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'capture_swarm_dpi.js')
DEFAULT_PROCESS = 'Turtle Beach Swarm II.exe'


def find_pid(name):
    try:
        for p in frida.get_local_device().enumerate_processes():
            if p.name.lower() == name.lower():
                return p.pid
    except Exception as e:
        _common.say('process enumerate failed: %s' % e)
    return None


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument('--process', default=DEFAULT_PROCESS, help='process that loads KONE_XP_AIR.dll')
    args = ap.parse_args()

    pid = find_pid(args.process)
    if pid is None:
        _common.say('%s is not running. Open Swarm II first (or try --process "Turtle Beach Device Service.exe").'
                    % args.process)
        return 1
    _common.say('Attaching to %s (pid %d)...' % (args.process, pid))
    try:
        session = frida.get_local_device().attach(pid)
    except Exception as e:
        _common.say('could not attach: %s\nRun this terminal as Administrator and try again.' % e)
        return 1

    count = {'n': 0}

    def on_message(msg, data):
        if msg.get('type') == 'send':
            p = msg.get('payload') or {}
            if 'line' in p:
                _common.say('[%s]' % p['line'])
            else:
                count['n'] += 1
                _common.say('%s  %s' % (p.get('dir', '??'), p.get('data', '')))
        elif msg.get('type') == 'error':
            _common.say('SCRIPT ERROR: %s' % (msg.get('description') or msg))

    with open(JS, encoding='utf-8') as f:
        script = session.create_script(f.read())
    script.on('message', on_message)
    script.load()

    _common.say('\n=== Now change the DPI in Swarm II (e.g. 800 -> 3000) and click apply. ===')
    try:
        input('Press ENTER here when done...\n')
    except (EOFError, KeyboardInterrupt):
        pass
    _common.say('Captured %d frames. Send back the SEND/GET lines from the DPI change.' % count['n'])
    try:
        session.detach()
    except Exception:
        pass
    return 0


if __name__ == '__main__':
    sys.exit(main())
